### DISCUSSION TITLE
Microtype + lua-widow-control

### DISCUSSION THREAD

**Discussion by hpvd** [gucci-on-fleek/lua-widow-control#15] [General]
from doc:

> ...If you look closely, you can see that this stretched the interword spaces. This stretch­
> ing is noticeable when typesetting in a narrow text block, but it becomes nearly
> imperceptible with larger widths.

This works great as one can see in the demo Table 1 in the doc!

As said, "This stretching is noticeable". Noticeable means the visual quality may be degraded.

In https://github.com/gucci-on-fleek/lua-widow-control/issues/3#issuecomment-1029618675 is documented, that the microtype package is already involved in optimizing the appearance.

=> so wouldn't it be possible to **solely** rely on microtype package to set the text to reach the desired length (to prevent widows and orphans)?

When doing this, the resulting visual appearance of the text should be further improved since the microtype package not only modifies spaces between the word but also within the words.

One idea would be, to locally widen the allowed ranges of all microtypes working fields as needed (more tight or more wide as normally allowed)

=> what do you think?

**Comment by gucci-on-fleek** [comment]
A few thoughts here:

## Stylistic Considerations

### Maximum Stretch/Shrink
Microtype by default sets `stretch`/`shrink` to `20`, which means that each glyph can expand/contract by 2%. 

In "Micro-typographic extensions to the TeX typesetting system", which is what introduced glyph stretching/shrinking to TeX, [Hàn Thế Thành writes](https://tug.org/TUGboat/tb21-4/tb69thanh.pdf#page=73):
> According to the experiments, the reasonable limit of font expansion (font stretch as well as font shrink) is about 20 for Computer Modern and Type 1 fonts. […] In general, **the variance in character widths must be limited to about ±2%.**

In "Hints and caveats" [in the Microtype documentation](https://texdoc.org/serve/microtype/0#section.11), we have:
> Don’t use too large a value for expansion. Font expansion is a feature that is supposed to enhance the typographic quality of your document by producing a more uniform greyness of the text block (and potentially reducing the number of necessary hyphenations). When expanding or shrinking a font too much, the effect will be turned into the opposite. **Expanding the fonts by more than 2%, i.e., setting a stretch limit of more than 20, should be justified by a typographically trained eye.**

And in "The Elements of Typographic Style", Robert Bringhurst writes:
> And the best computer justification now relies on microscopic adjustments to the space between and within the letters as well as the space between the words. In this book, for example, the justification engine has been permitted to vary the intercharacter spacing by ±3% and **to adjust the width of individual glyphs by ±2%**. The bulk of the work is still done by adjusting the spaces between words, but there are more letters than spaces in these lines. Tiny adjustments to spaces within and between the letters therefore go a long way toward creating a page of even color and texture.

Pretty much every source that I've read says that you probably shouldn't increase the glyph expansion beyond 2%. It is probably safe to increase it a little further sometimes, but this should be done manually—not automatically.

Also, it's probably not a good idea to change the expansion settings for different paragraphs in a document. It would be better to set larger `stretch`/`shrink` values _globally_—TeX always tries to minimize the expansion/contraction of the glyphs, so the glyphs are much more likely to be stretched in an expanded paragraph.

### Demo

This test file here shows what happens if you disable word space stretching, and only justify a paragraph with glyph stretching. Compile with `luatex` (not `lualatex`).
```tex
%!TeX program = luatex

\hsize=345pt
\hoffset=62pt
\vsize=550pt
\voffset=56pt

\def\setfontsettings{
    \spaceskip=\fontdimen2\font plus 0.05\fontdimen3\font minus 0.05\fontdimen4\font\relax
    \xspaceskip=\fontdimen2\font plus 0.05\fontdimen3\font minus 0.05\fontdimen4\font\relax
    \expandglyphsinfont\the\font 50 50 1
    \adjustspacing=2
}

\input plipsum

\setfontsettings
\lipsum{1-100}
\vfill\eject

\font\myfont=cmss10\myfont
\setfontsettings
\lipsum{1-100}
\vfill\eject

\bye
```
[issue-4.pdf](https://github.com/gucci-on-fleek/lua-widow-control/files/8009522/issue-4.pdf)

To eliminate all the `overfull hbox`es, I had to set the `strech`/`shrink` values to `50`, or 5%. This is only double the default, but it's enough to make the demo look pretty bad with Computer Modern. Now, this method works pretty well with sans serif fonts, but (La)TeX documents that use sans fonts are pretty rare.

And one more demo here:
```tex
\parindent=0pt

\hoffset=-1in
\voffset=-1in

\hsize=520pt
\pagewidth=\hsize

\vsize=28\baselineskip
\pageheight=\vsize

\raggedbottom

\def\testsentence{This is a short test sentence. And this is a longer test sentence.}

\input pgffor

\newdimen\onepoint
\onepoint=1pt

\def\setexpand#1{
    \font\newfont=cmr10\newfont
    \expandglyphsinfont\newfont #1 #1 5
    \adjustspacing=2
}

\foreach \y in {5, 10, ..., 50, 60, 70, ..., 200, 300, 400, ..., 1000}{
    \setexpand{\y}
    \hskip 3em \rlap{{\bf Expansion Factor: \y/1000}}%
    \hskip 278.61143pt\smash{\vrule height 2\baselineskip depth 28\baselineskip width 0.5pt}

    \foreach \x in {-60, -50, ..., 200}{
        \parfillskip=\dimexpr\hsize - 278.61143pt - 3em - \x\onepoint \relax
        \leavevmode\hbox to 3em {\x}\testsentence\par
    }
    \vfill\eject
}

\bye
```

https://user-images.githubusercontent.com/49086429/152669637-0725d2fb-1f2a-41d2-885c-69117e5be4f9.mp4

The left column shows how many points away from its natural length the paragraph has been stretched. You can see the behaviour as the expansion factor increases. From watching this, I'd say that 4–5% is pretty much the limit before the glyphs start to look distorted.

So if 2% expansion is the default, and 4–5% is enough to look bad, there's not really much room for lua-widow-control to locally increase the expansion. If you wanted more expansion, you'd be best to set it globally.

## Technical Considerations

Nevertheless, I tried to implement this in lua-widow-control, and I don't think that it's possible. 

In `luatex`, each font can only have one stretch limit¹. This means that you need to copy a (virtual) font to change its stretch limit. However, copying a font is non-trivial in `luatex`, so this is gets pretty tricky. I could maybe use Microtype's internals to do this, but even that would get pretty complicated.

Also, each paragraph can only have one stretch limit². This means that to change the stretch for a paragraph, you'd need to change the stretch for all the fonts in it (regular, bold, italic, and math). 

But even if you do that, `luatex` still gives you the same error. Lua-widow-control only "intercepts" the paragraph _after_ TeX has already processed it. It seems like the font expansion settings have already been locked in at this point, so there's not much that you can do.

Now, I could probably [fix this in ConTeXt/LMTX](https://mailman.ntg.nl/pipermail/ntg-context/2019/096406.html), but very few people use ConTeXt, so it would be a pretty niche feature.

So unfortunately, I don't think that this is possible to implement. That being said, you can submit a PR if you do somehow figure out how to implement this.

¹`! error: (font expansion): font has been expanded with different stretch limit.`
²`! error: (font expansion): using fonts with different limit of expansion in one paragraph is not allowed.`

**Comment by hpvd** [comment]
wow many thanks for doing this pretty advanced examination
and stunning visualisation!

Just to clearify: was always thinking about space modification within AND between words.

will get back with some thoughts within next days.

**Comment by hpvd** [comment]
Since this lua widow control and the microtype package still looks like an automation dream-team to me,
I did some "promotion" for this package and the remaining question of this issue: https://github.com/schlcht/microtype/issues/9

**Comment by schlcht** [comment]
Thanks, @hpvd for drawing my attention to this very interesting package. I basically agree with @gucci-on-fleek's thorough analysis above though. The gain is probably limited, as the expansion limits shouldn't be increased too much, but first and foremost I think it would be very hard to implement. 

Just some random remarks from my side: 

- I would not recommend to increase the expansion settings _globally_: because there is no penalty attached to expansion, tex would often unnecessarily "over-expand".
- Is there a particular reason why you make paragraphs only longer, not shorter? Having both options could provide more leeway.
- For the documentation example, I would switch on frenchspacing, as the larger spaces after punctuation tend to ruin the greyness value.
- Testing the package, I was missing some debug information, which would be nice to have as an option ...

**Comment by gucci-on-fleek** [comment]
@schlcht Thanks for the feedback! I'll respond to each point below:

> there is no penalty attached to expansion

Huh, I had no idea. I had just assumed that font expansion accumulated badness just like the stretch-ratio-cubed calculation used for glue. 

> For the documentation example, I would switch on frenchspacing, as the larger spaces after punctuation tend to ruin the greyness value.

Yep, definitely a good idea. Thanks!

> Testing the package, I was missing some debug information, which would be nice to have as an option ...

I probably should add in some more debugging information. That will certainly make things much easier for me. I'll add this in in one of the next few releases or so

---

> Is there a particular reason why you make paragraphs only longer, not shorter? 

I tried shortening paragraphs in some _very_ early lua-widow-control versions, but I took it out in 582b007a429be13e1187d6f33664acf70d8d99d8. The problem is that there really aren't that many paragraphs in a document that you can easily shrink. If you run this (messy) code:
```latex
\documentclass{article}
\usepackage{luacode}
\usepackage{kantlipsum}

\emergencystretch=3em

% Adapted from https://tex.stackexchange.com/questions/372062/paragraph-callback-to-help-with-widows-orphans-hand-tuning
\begin{luacode*}
    only_expand = 0
    only_shrink = 0
    only_emerg_expand = 0
    expand_and_shrink = 0
    emerg_expand_and_shrink = 0
    paragraphs = 0

    local pre_linebreak_test_looseness = function (head)
        local n, i = tex.linebreak(node.copy_list(head))
        local nP1s, iP1s = tex.linebreak(node.copy_list(head), {looseness=tex.looseness+1})

        local nP1, iP1
        if iP1s.prevgraf > i.prevgraf then
            nP1, iP1 = tex.linebreak(node.copy_list(head), {looseness=tex.looseness+1, emergencystretch=0})
        else
            nP1, iP1 = n, i
        end

        local nM1, iM1 = tex.linebreak(node.copy_list(head), {looseness=tex.looseness-1})

        if iP1.prevgraf > i.prevgraf and iM1.prevgraf < i.prevgraf then
            expand_and_shrink = expand_and_shrink + 1
        elseif iP1s.prevgraf > i.prevgraf and iM1.prevgraf < i.prevgraf then
            emerg_expand_and_shrink = emerg_expand_and_shrink + 1
        elseif iP1.prevgraf > i.prevgraf then
            only_expand = only_expand + 1
        elseif iP1s.prevgraf > i.prevgraf then
            only_emerg_expand = only_emerg_expand + 1
        elseif iM1.prevgraf < i.prevgraf then
            only_shrink = only_shrink + 1
        end
        paragraphs = paragraphs + 1

        return true
    end

    luatexbase.add_to_callback("pre_linebreak_filter", pre_linebreak_test_looseness, "pre_linebreak_test_looseness")
\end{luacode*}

\begin{document}
    \kant[1-164]

    \begin{luacode*}
        print("Expansion Only", only_expand)
        print("Shrinking Only", only_shrink)
        print("Emerg. Expansion Only", only_emerg_expand)
        print("Expand and Shrink", expand_and_shrink)
        print("Emerg. Expand and Shrink", emerg_expand_and_shrink)
        print("Total Paragraphs", paragraphs)
        print("Total expand/shrink possible", only_expand + only_shrink + only_emerg_expand + expand_and_shrink + emerg_expand_and_shrink)
    \end{luacode*}
\end{document}
```
Then you get:
> |Type|Count|
> |---|---|
> |Expansion Only|100|
> |Shrinking Only|3|
> |Emerg. Expansion Only|26|
> |Expand and Shrink|4|
> |Emerg. Expand and Shrink|31|
> |Total Paragraphs|164|
> |Total expand/shrink possible|164|

When I run the same code on [TeX by Topic](https://www.ctan.org/pkg/texbytopic), I get:
> |Type|Count|
> |---|---|
> |Expansion Only|1578|
> |Shrinking Only|39|
> |Emerg. Expansion Only|556|
> |Expand and Shrink|0|
> |Emerg. Expand and Shrink|1|
> |Total Paragraphs|7270|
> |Total expand/shrink possible|2174|

So it seems like 20–60 % of paragraphs are only-expandable, but only 0.5–2 % of paragraphs are only-shrinkable. In the above examples, there are around 3–25 paragraphs per page, so it's pretty rare to find a page where there is a paragraph that you can shorten but none that you can lengthen. 

Shortening _and_ lengthening paragraphs adds a fair bit of complexity to the Lua code, so it's probably not worth the hassle to add it back in. Nevertheless, I'll experiment with it a bit and maybe add it in as a feature if it's not too hard to implement.
